README.md
# Soil bacterial populations are shaped by recombination and gene-specific selection across a grassland meadow
Reproducible scripts and notebooks for 2019 paper on population genetics in metagenomes

## Structure

*inStrain_lite* - A `pip` installable python program that will call SNPs and calculate nucleotide diversity, linkage, coverage+breadth, and Fst.
Meant to be reusable for other projects, eventually will be published as its own independent repository. 

*Data* - contains some of the usable data files for this paper.

*Notebooks* - notebooks used to generate each figure in this paper.

## Reproducibility

### Genome dereplication
dRep was run on all 10,538 genome bins obtained, with the secondary clustering threshold `-sa 0.97`.
CheckM genome information for all bins is shown in `genomeInfo.csv`.

### Pan-genome decontamination 
The representative genomes used in this study were decontaminated using the assembled replicate genomes for each species population. Roary was run with default settings on the set of genomes for each species. Contigs with at least 50% of their genes found in no more than 25% of each genome set were discarded as potential contamination. 

### Read filtering

Read filtering is performed using the custom script `filter_reads.py`. Briefly, a bowtie2 index is created using representative members for all 664 dereplicated genomes. Reads were then mapped to this index using bowtie2 default settings with the expected insert distance `-X 1000`. The resulting BAM files are then filtered for reads that meet the following criteria:
(1) Both read pairs map to the same scaffold within 1500 bp of each other (end to end insert size), with a percent identity of 96%.
(2) At least one of the read pairs has a mapq score > 1, e.g. this is a uniquely best mapping for this read (and therefore probably for this pair) in the index.

Reads in this dataset are on average ~200 bp. (HiSeq 2500 sequencing that aimed for 2x250 bp reads, but the trailing ends of the reads were often low quality and so they were usually trimmed to ~200 bp. 

### Meadow-wide population profiling.

A BAM file for each species containing all filtered reads assigned to that species from the meadow was created using `combine_filter_bams.py` in `./meadow_wide/`.  The commands run are in `./meadow_wide/run_combine.sh`. These BAMs were then used to analyze total nucleotide diversity and SNP linkage across the meadow. These BAMs were then profiled with the `inStrain_lite` script with the following parameters:
```
inStrain_lite -p 48 -s 30 -c 0.96 ../meadow_wide/14_0903_02_20cm_Proteobacteria_56_68_14_filtered_sort.bam  ../representative_genomes
/14_0903_02_20cm_Proteobacteria_56_68_14.fasta
```

All inStrain_lite commands run are in the `./meadow_wide/run_inStrain.sh`. 

### Per-sample population profiling

Each individual sample was profiled separately for the purpose of calculating nucleotide diversity per sample / plot, and Fst between plots. The inStrain commands run are in `./strains_data/run_instrain.sh` and take the form: 
```
inStrain_lite -p 48 -s 30 -c 0.96 --min_breadth_cov 0.5,5 ../bams/all_14_0903_02_20cm_sorted.bam ../representative_genomes/14_0903_02_20cm_Proteobacteria_56_68_14.fasta
```
586 out of 1140 total (19 genomes*60 samples) genome+sample pairs passed the minimum requirement of at least 5x filtered read coverage across at least 50% of the representative genome to be included in these downstream analyses. 

Data from samples were then aggregated into various groupings (by replicate, plot, block, and all) using `./merged_samples/make_merge_commands.py` which generates commands for all of the possible combinations to run `./inStrain_lite/combine_samples.py`. Merged samples were used for Figure 3 and for FST data (merged by block).

### Gene statistics

Genes were called on all representative genomes with `prodigal` - output is in `./representative_genomes/fna/*`. The script `gene_statistics.py` was run on each of the per-sample population profiles, each aggregated profile, and the meadow-wide profiles (commands found in scripts called `run_gene_commands.sh` in each respective `./data/` directory).

### Organism abundances

Organism relative abundances were taken by counting the total number of reads that mapped to each genome in each sample (`./data_tables/bin_reads.txt`) and dividing by the total number of reads per sample (`bin_total_reads.tsv`). 


inStrain_lite/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project (attempts to) adhere to [Semantic Versioning](http://semver.org/).

## [0.3.0] - 2019-05-15
- Now calculates D', and normalized versions of D' and R2
- min_snp is now 
- Has the new gene_statistics.py and combine_samples.py scripts

## [0.2.8] - 2019-05-09
- fixed varbase = refbase error when there's a tie for most abundant variant
- made default output prefix be the fasta file prefix
- now generate scafold counts numpy array which show total variant counts for all positions by scaffold

## [0.2.7] - 2019-04-10
### Changed
- Change to how scaff2pair2info is handled 

## [0.2.6] - 2019-04-09
### Changed
- Minor change to the pickle protocol and stuff

## [0.2.5] - 2019-04-07
### Changed
- Allow changing filtering criteria, and some basic testing to make sure it works
- Will convert .sams to .bams and index for you

## [0.2.4] - 2019-03-29
### Changed
- Add a little bit of logging and efficiency into read report

## [0.2.3] - 2019-03-26
### Changed
- multiprocessing of scaffolds re-factored to be better
- now produces readable dataframe outputs
- make a nice read report
- add a --read_report option for filter_reads

## [0.2.2] - 2019-03-26
### Changed
- paired read filtering is now multi-processed

## [0.2.1] - 2019-03-26
### Changed
- argparse now displays the version number and defaults
- runs Matts way by default

## [0.2.0] - 2019-03-26
### Fixed
- implemented logging module
- made read filtering into a saved table
- implemented and tested "filter_reads.get_paired_reads"
- multiprocessing of SNP calling implemented
- setup.py is made; can now install itself
- changed the quoting on CC's table outputs to default
- changed "level" to "filter_cutoff"
- changed "min_coverage" to "min_cov"
- account for read overlap when calculating read length (half code is there, but reverted)
- changed to "no_filter" stepper
- clonality is now called the same with CC and Matt's versions
- min_cov is called with >= now

## [0.1.2] - 2019-03-12
- Stepper is now all

## [0.1.1] - 2019-03-12
- Tests now work

## [0.0.4] - 2018-06-25
- Adding some of the graph stuff and functions. The Python
is really slow, in the future it should call node2vec C++ library.

## [0.0.2] - 2018-06-21
### Fixed
- Test suite now runs on it's own and makes sure it succeeds

## [0.0.2] - 2018-06-21
### Fixed
- Test suite produces a command
- Program doesn't crash (lotsa small bug fixes)

## [0.0.1] - 2018-06-21
### Added
- Changelog and versioning is born
- Test suite is born

### Fixed
- Dumb print error is fixed
- Import statement is pretty :)


inStrain_lite/inStrain_lite/__init__.py
#!/usr/bin/env python
# inStrain
# from ._version import __version__
import os
import csv
import sys
import pysam
import pickle
import logging
import argparse
import traceback
import itertools
import collections
import numpy as np
import pandas as pd
import networkx as nx
from tqdm import tqdm
from Bio import SeqIO
import concurrent.futures
from subprocess import call
from itertools import permutations
from collections import defaultdict
## local imports
from inStrain_lite.utilities import *
from inStrain_lite.filter_reads import filter_paired_reads
from inStrain_lite.linkage import build_SNP_linkage_network
from inStrain_lite.linkage import calculate_ld

P2C = {'A':0, 'C':1, 'T':2, 'G':3}
C2P = {0:'A', 1:'C', 2:'T', 3:'G'}

def call_snv_site(counts, min_cov=5, min_freq=0.05, model=None):
    '''
    Determines whether a site has a variant based on its nucleotide count frequencies.
    Return:
        Base if SNP
        -1 if not SNP
        None if unCounted
    '''
    if model: #alexcc - so you can call this function from outside of the file
        model_to_use = model
    else:
        model_to_use = null_model

    total = sum(counts)
    if total >= min_cov:
        i = 0
        for c in counts:
            if c >= model_to_use[total] and float(c) / total >= min_freq:
                i += 1
        if i > 1:
            return C2P[np.argmax(counts)]
        else:
            return None
    else:
        return None


###############################################################
######## START MAIN SNP PROFILE CLASS OBJECT ##################
###############################################################

class SNPprofile():

    def __init__(self):

        # parameters
        self.output = None
        self.fasta_db = None
        self.scaffold_list = []
        self.counts_table = None
        self.coverage_table = None
        self.snp_table = None
        self.linkage_table = None

        # snp calling model 
        null_loc = os.path.dirname(__file__) + '/helper_files/combined_null1000000.txt'
        global null_model
        null_model = generate_snp_model(null_loc)

    def save(self):
        if self.coverage_table is not None:
            self.coverage_table.to_csv(self.output + '.scaffold_info.tsv', index=False, sep='\t')
        if self.snp_table is not None:
            self.snp_table.to_csv(self.output + '.SNVs.tsv', index=False, sep='\t')
        if self.linkage_table is not None:
            self.linkage_table.to_csv(self.output + '.linkage.tsv', index=False, sep='\t')
        f = open(self.output + ".data", 'wb')
        pickle.dump(self.__dict__, f, 2)
        f.close()

    def load(self, name):
        self.__dict__.clear()
        f = open(name + ".data", 'rb')
        tmp_dict = pickle.load(f)
        f.close()
        self.__dict__.update(tmp_dict)

    def load_fasta(self, args):
        '''
        Load the FASTA sequences to be profiled.
        Return a table listing scaffold name, start, end
        '''
        # PROFILE ALL SCAFFOLDS IN THE .FASTA FILE
        scaff2sequence = SeqIO.to_dict(SeqIO.parse(args.fasta, "fasta")) # set up .fasta file
        s2l = {s:len(scaff2sequence[s]) for s in list(scaff2sequence.keys())} # Get scaffold2length
        Fdb = pd.DataFrame(list(s2l.items()), columns=['scaffold', 'end'])
        Fdb['start'] = 0
        self.fasta_db = Fdb
        return True

    def profile(self, bam, filtered_reads, min_freq=0.05, min_ani=0.97, min_mapq=2, min_insert=50, max_insert=1500, min_ld = 30, min_breadth_cov=[0,0], threads=6):
        ''' Profiles one or multiple scaffolds'''
        args = {"min_freq": min_freq, "min_ani": min_ani, "min_mapq":min_mapq, "min_insert":min_insert, "max_insert":max_insert, 'min_ld': min_ld}
        with concurrent.futures.ProcessPoolExecutor(max_workers=threads) as executor:
            Sprofiles = [s for s in tqdm(executor.map(self.profile_scaffold,
                                            self.iterate_commands(self.fasta_db, bam, filtered_reads, args)),
                                            desc='Profiling scaffolds: ',
                                            total=len(self.fasta_db['scaffold'].unique()))]
        
        logging.info("Finalizing and saving")

        counts_table = list()
        counts_array = np.zeros(shape=(0, 4), dtype=int)
        coverage_breadth_table = defaultdict(list)
        i = 0
        for Sprof in Sprofiles:

            # create counts table
            counts_table.append(Sprof[0])
            self.scaffold_list.append(Sprof[3])
            counts_array = np.append(counts_array, Sprof[0], axis=0)
            coverage_breadth_table['scaffold'].append(Sprof[3])
            coverage_breadth_table['coverage'].append(np.mean(np.sum(Sprof[0], axis=1)))
            coverage_breadth_table['breadth'].append((np.sum( Sprof[0], axis=1) > 0).sum() / float( Sprof[0].shape[0]))

            # create snp / linkage tables
            if i == 0:
                snp_table = Sprof[1]
                linkage_table = Sprof[2]
            else:
                snp_table = pd.concat([snp_table, Sprof[1]])
                linkage_table = pd.concat([linkage_table, Sprof[2]])
            i += 1   

        coverage_breadth_table = pd.DataFrame(coverage_breadth_table)
        total_coverage = np.mean(np.sum(counts_array, axis=1))
        total_breadth = (np.sum(counts_array, axis=1) > 0).sum() / float(counts_array.shape[0])
        total_min_cov_breadth = (np.sum(counts_array, axis=1) >= min_breadth_cov[1]).sum() / float(counts_array.shape[0])

        logging.info("Mean coverage\t" + str(total_coverage))
        logging.info("Mean breadth\t" + str(total_breadth))
        logging.info("Breadth above minimum coverage\t" + str(total_min_cov_breadth))

        if total_min_cov_breadth >= min_breadth_cov[0]:
            logging.info("Passes breadth-cov filter. Will output tables")
            self.counts_table = counts_table
            self.coverage_table = coverage_breadth_table
            self.snp_table = snp_table
            self.linkage_table = linkage_table
            self.save()
        else:
            logging.info("NOTE: does NOT pass breadth-cov filter. Will not output tables")

    def iterate_commands(self, Fdb, bam, filtered_reads, args):
        '''
        Make and iterate profiling commands
        Doing it in this way makes it use way less RAM
        '''
        for index, row in Fdb.iterrows():
            cmd = {"scaffold": row['scaffold'], 'scaf_length': row['end'], "bam": bam, "filtered_reads": filtered_reads, "args":args}
            yield cmd

    def profile_scaffold(self, args):
        snp_table = defaultdict(list)
        read2snps = defaultdict(list)
        samfile = pysam.AlignmentFile(args['bam'])
        try:
            iter = samfile.pileup(args['scaffold'], truncate = True, max_depth=100000,
                                    stepper = 'nofilter', compute_baq= True,
                                    ignore_orphans = True, ignore_overlaps = True,
                                    min_base_quality = 30)
        except ValueError:
            logging.error("ERROR " + args['scaffold'] + " is not in the BAM file.")
            sys.exit(1)

        pileup_counts = np.zeros(shape=(args['scaf_length'], 4), dtype=int) # Holds all pileup counts - alexcc 5/9/2019
        for pileupcolumn in iter:
            counts = get_base_counts(pileupcolumn, args['filtered_reads'])

            counts_sum = np.sum(counts)

            pileup_counts[pileupcolumn.pos,] = counts
            snp = call_snv_site(counts, min_freq = args['args']['min_freq'])
            if snp: # means this is a SNP
                snp, varbase = major_minor_allele(counts)
                # Add to SNP table
                snp_table['scaffold'].append(args['scaffold'])
                snp_table['position'].append(pileupcolumn.pos)
                for b, c in zip(['A', 'C', 'T', 'G'], counts):
                    snp_table[b].append(c)
                snp_table['conBase'].append(snp)
                snp_table['varBase'].append(varbase)
                snp_table['baseCoverage'].append(np.sum(counts))
                # Add to read2snps.
                for pileupread in pileupcolumn.pileups:
                    if not pileupread.is_del and not pileupread.is_refskip and pileupread.alignment.query_name in args['filtered_reads']:
                        read2snps[pileupread.alignment.query_name].append( (pileupcolumn.pos, pileupread.alignment.query_sequence[pileupread.query_position]) )
        samfile.close()

        snp_table = pd.DataFrame(snp_table)
        snp_net, linkage_net = build_SNP_linkage_network(read2snps)
        linkage_table = calculate_ld(snp_net, linkage_net, args['scaffold'], pileup_counts, args['args']['min_ld'])
        return pileup_counts, snp_table, linkage_table, args['scaffold']
 

###############################################################
######## SCRIPT CONTROLLER CLASS  #############################
###############################################################

class Controller():
    '''
    Main controller of the program
    '''

    def main(self, sys_args):
        '''The main method when run on the command line'''
    
        strains = SNPprofile()

        args = parse_arguments(sys_args)
        if not args.output:
            strains.output = args.fasta.split("/")[-1].split(".")[0] + ":" + args.bam.split("/")[-1].split(".")[0]
        else:
            strains.output = args.output

        setup_logger(strains.output + '.log')
        args.min_breadth_cov = [float(i) for i in args.min_breadth_cov.split(",")[:2]]
        
        strains.load_fasta(args)

        filtered_reads = filter_paired_reads(args.bam, args.fasta, args.min_ani, args.min_mapq, args.min_insert, args.max_insert)

        strains.profile(args.bam, filtered_reads, args.min_freq, args.min_ani, args.min_mapq, args.min_insert, args.max_insert, args.min_ld, args.min_breadth_cov, args.threads)


def parse_arguments(args):
    parser = argparse.ArgumentParser(description="inStrain",
             formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    # Required positional arguments
    parser.add_argument("bam", help="Sorted .bam file")
    parser.add_argument("fasta", help="Fasta file the bam is mapped to")

    # Optional arguments
    parser.add_argument("-o", "--output", action="store", default=None, \
        help='Output prefix')
    parser.add_argument("-p", "--threads", action="store", default=6, type=int, \
        help='Threads to use for multiprocessing')
    parser.add_argument("-s", "--min_ld", action="store", default=20, \
        help='Minimum number of reads connecting two segregating sites to calculate LD between them. (>=)')
    parser.add_argument("-f", "--min_freq", action="store", default=0.05, \
        help='Minimum SNP frequency to confirm a SNP (both this AND the 10^-6 FDR SNP count cutoff must be true).')

    # Read filtering cutoffs
    parser.add_argument("-c", "--min_ani", action="store", default=0.95, type=float, \
        help='Minimum required percent identity of read pairs to consensus. (>=)')
    parser.add_argument("--min_mapq", action="store", default=2, type=int,\
        help='Minimum mapq score of EITHER read in a pair to use that pair. (>=) ')
    parser.add_argument("--min_insert", action="store", default=50, type=int,\
        help='Minimum insert size of two reads - default is 50 bp. True insert size includes read lengths: if two reads are 50bp each and overlap completely, their insert will be 50. (>=)')
    parser.add_argument("--max_insert", action="store", default=1500, type=float, \
        help='Maximum insert size between two reads - default is 1500 bp (<=)')
    parser.add_argument("--min_breadth_cov", action="store", default="0,0", type=str,\
        help='OPTIONAL: comma separated min breadth and a specific coverage required to generate output. E.g., 0.50,5 will require 50 percent of the genome to be at least 5x coverage to generate tables')

    # Parse
    if (len(args) == 0 or args[0] == '-h' or args[0] == '--help'):
        parser.print_help()
        sys.exit(0)
    else:
        return parser.parse_args(args)

if __name__ == '__main__':
    Controller().main(sys.argv[1:])
